chain-impl-mockchain/src/leadership/bft.rs

```diff
@@ -0,0 +1,147 @@
+use crate::block::SignedBlock;
+use crate::key::PublicKey;
+use crate::leadership::IsLeading;
+use crate::update::{BftSelectionDiff, ValueDiff};
+
+use chain_core::property::{self, LeaderSelection};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct BftRoundRobinIndex(usize);
+
+/// The BFT Leader selection is based on a round robin of the expected leaders
+#[derive(Debug)]
+pub struct BftLeaderSelection<LeaderId> {
+    my: Option<BftRoundRobinIndex>,
+    leaders: Vec<LeaderId>,
+
+    current_leader: LeaderId,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    BlockHasInvalidLeader(PublicKey, PublicKey),
+    BlockSignatureIsInvalid,
+    UpdateHasInvalidCurrentLeader(PublicKey, PublicKey),
+}
+
+impl<LeaderId: Eq + Clone> BftLeaderSelection<LeaderId> {
+    /// Create a new BFT leadership
+    pub fn new(me: LeaderId, leaders: Vec<LeaderId>) -> Option<Self> {
+        if leaders.len() == 0 {
+            return None;
+        }
+
+        let pos = leaders
+            .iter()
+            .position(|x| x == &me)
+            .map(BftRoundRobinIndex);
+        let current_leader = leaders[0].clone();
+        Some(BftLeaderSelection {
+            my: pos,
+            leaders: leaders,
+            current_leader: current_leader,
+        })
+    }
+
+    #[inline]
+    pub fn number_of_leaders(&self) -> usize {
+        self.leaders.len()
+    }
+
+    #[inline]
+    fn offset(&self, block_number: u64) -> BftRoundRobinIndex {
+        let max = self.number_of_leaders() as u64;
+        BftRoundRobinIndex((block_number % max) as usize)
+    }
+
+    /// get the party leader id elected for a given slot
+    #[inline]
+    pub fn get_leader_at(&self, slotid: u64) -> &LeaderId {
+        let BftRoundRobinIndex(ofs) = self.offset(slotid);
+        &self.leaders[ofs]
+    }
+
+    /// check if this party is elected for a given slot
+    #[inline]
+    pub fn am_leader_at(&self, slotid: u64) -> IsLeading {
```

Comment:
Is this type supposed to represent both the set of leaders and the selected leader? The name also confusingly differs from `LeaderSelection::is_leader_at`.

Reply:
good point, and it overlaps with the method exposed by the `LeaderSelection` trait. I'll make it private and change the name too.